File: grep_recursive_tool.py
import re
from pathlib import Path
from check_file_permission import check_file_permission

def grep_recursive(dir_path, pattern, config, regex=False):
    if not check_file_permission(dir_path, "read", config):
        raise PermissionError(f"Read access denied for {dir_path}")
    dir_path = Path(dir_path).resolve()
    result = []
    for file_path in dir_path.rglob("*"):
        if file_path.is_file() and check_file_permission(str(file_path), "read", config):
            with open(file_path, "r", encoding="utf-8") as f:
                for i, line in enumerate(f, 1):
                    if regex and re.search(pattern, line):
                        result.append(f"{file_path}:{i}:{line.rstrip()}")
                    elif pattern in line:
                        result.append(f"{file_path}:{i}:{line.rstrip()}")
    return "\n".join(result) if result else "No matches found"

File: ts-tools/grep_recursive_tool.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

async function* walk(dir: string): AsyncGenerator<string> {
  for await (const d of await fs.promises.opendir(dir)) {
    const entry = path.join(dir, d.name);
    if (d.isDirectory()) yield* walk(entry);
    else if (d.isFile()) yield entry;
  }
}

export async function grepRecursive(
  dirPath: string,
  pattern: string,
  config: any,
  regex: boolean = false,
): Promise<string> {
  await checkFilePermission(dirPath, 'read', config);
  const resolvedPath = path.resolve(dirPath);
  const result: string[] = [];
  const re = regex ? new RegExp(pattern) : null;

  for await (const p of walk(resolvedPath)) {
    await checkFilePermission(p, 'read', config);
    const data = await fs.promises.readFile(p, 'utf-8');
    const lines = data.split('\n');
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (re) {
        if (re.test(line)) {
          result.push(`${p}:${i + 1}:${line}`);
        }
      } else {
        if (line.includes(pattern)) {
          result.push(`${p}:${i + 1}:${line}`);
        }
      }
    }
  }
  return result.length > 0 ? result.join('\n') : 'No matches found';
}
